src/net.cpp

```diff
@@ -2039,7 +2039,29 @@ void CConnman::DisconnectNodes()
     // m_reconnections_mutex while holding m_nodes_mutex.
     decltype(m_reconnections) reconnections_to_add;
 
-    {
+    // Quick check without exclusive lock to avoid unnecessary locking
+    // Note: This is a best-effort optimization. If network is active and no nodes
+    // are marked for disconnect, we can skip the expensive exclusive lock.
+    // If network is inactive or nodes need cleanup, we must take the lock.
+    bool has_to_disconnect = false;
+    if (fNetworkActive) {
+        {
+            READ_LOCK(m_nodes_mutex);
+            for (CNode* pnode : m_nodes) {
+                if (pnode->fDisconnect) {
+                    has_to_disconnect = true;
+                    break;
+                }
+            }
+        }
+        // Only return early if network is active AND no nodes need disconnection
+        // AND no disconnected nodes need cleanup (checked below)
+        if (!has_to_disconnect && m_nodes_disconnected.empty()) {
+            return;
+        }
+    }
+
+    if (has_to_disconnect || !fNetworkActive) {
```

Comment:
what if status "network active" has been changed from 'true' to 'false' - does it mean that nodes won't ever disconnected?

Reply:
ah, I read this condition as:
```
if (has_to_disconnect || fNetworkActive) {
```

my bad